File: Medium_Level_using_Python/Arrays_and_Strings/15. 3Sum.py
# Problem: 15. 3Sum
# Link: https://leetcode.com/problems/3sum/description/
# Time Complexity: O(n^2)
# Space Complexity: O(1)
# Topics: Array, Two Pointers, Sorting
# Solution:
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        triplets = []
        lenNums = len(nums)

        nums.sort()   # Sort the List

        for i in range (0, lenNums - 2):  # Iterate through the nums
            if i > 0 and nums[i] == nums[i - 1]:  # Skip the duplicates
                continue

            target = -nums[i]   # Since nums[i] + nums[j] + nums[k] == 0
            left = i + 1
            right = lenNums - 1

            while left < right:
                currSum = nums[left] + nums[right]

                if currSum == target:
                    triplets.append([nums[i], nums[left], nums[right]])

                    left += 1
                    right -= 1

                    while left < right and nums[left] == nums[left - 1]:  # Skip the duplicates
                        left += 1
                    while left < right and nums[right] == nums[right + 1]:  # Skip the duplicates
                        right -= 1

                elif currSum < target:
                    left += 1
                else:
                    right -= 1

        return triplets

File: Medium_Level_using_Cpp/Array_and_String/15. 3Sum.cpp
// Problem: 15. 3Sum
// Link: https://leetcode.com/problems/3sum/description/
// Time Complexity: O(n^2)
// Space Complexity: O(1)
// Topics: Array, Two Pointers, Sorting
// Solution: 
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> triplets;
        sort(nums.begin(), nums.end());
        int lenNums = nums.size();

        for(int i = 0; i < lenNums - 2; i++){
            if( i > 0 && nums[i] == nums[i - 1]){
                continue;
            }

            int target = -nums[i];
            int left = i + 1;
            int right = lenNums - 1;

            while (left < right){
                int currSum = nums[left] + nums[right];
                
                if (currSum == target){
                    triplets.push_back({nums[i], nums[left], nums[right]});

                    left++;
                    right--;

                    while(left < right && nums[left] == nums[left - 1]){
                        left++;
                    }
                    while(left < right && nums[right] == nums[right + 1]){
                        right--;
                    }
                }
                else if (currSum < target) {
                    left++;
                }
                else{
                    right--;
                }
            }   
        }
        return triplets;
    }
};